File: Tyche.Android/user/src/main/java/com/felipearpa/tyche/user/login/ui/LoginViewModel.kt
package com.felipearpa.tyche.user.login.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.core.network.NetworkException
import com.felipearpa.tyche.ui.UnknownLocalizedException
import com.felipearpa.tyche.ui.ViewState
import com.felipearpa.tyche.ui.network.toNetworkLocalizedException
import com.felipearpa.tyche.user.UserProfile
import com.felipearpa.tyche.user.login.application.LoginUseCase
import com.felipearpa.tyche.user.login.application.toLoginInput
import com.felipearpa.tyche.user.login.domain.LoginException
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class LoginViewModel @Inject constructor(private val loginUseCase: LoginUseCase) : ViewModel() {

    private val _state = MutableStateFlow<ViewState<UserProfile>>(ViewState.Initial)
    val state: Flow<ViewState<UserProfile>>
        get() = _state.asStateFlow()

    fun resetState() {
        viewModelScope.launch {
            _state.emit(ViewState.Initial)
        }
    }

    fun login(loginCredentialCredential: LoginCredentialModel) {
        viewModelScope.launch {
            _state.emit(ViewState.Loading)
            loginUseCase.execute(loginInput = loginCredentialCredential.toLoginInput())
                .onSuccess { userProfile -> _state.emit(ViewState.Success(userProfile)) }
                .onFailure { exception -> _state.emit(ViewState.Failure(exception.toLocalizedException())) }
        }
    }
}

private fun Throwable.toLocalizedException() =
    when (this) {
        is LoginException -> this.toLoginLocalizedException()
        is NetworkException -> this.toNetworkLocalizedException()
        else -> UnknownLocalizedException()
    }

File: Tyche.iOS/User/Sources/User/Login/UI/LoginViewModel.swift
import Foundation
import Core
import UI

public class LoginViewModel: ObservableObject {
    
    @Published @MainActor private(set) var state: ViewState<UserProfile> = .initial
    
    private let loginUseCase: LoginUseCase
    
    public init(loginUseCase: LoginUseCase) {
        self.loginUseCase = loginUseCase
    }
    
    func resetState() {
        Task {
            await MainActor.run {
                state = .initial
            }
        }
    }
    
    func login(loginCredential: LoginCredentialModel) {
        Task {
            await MainActor.run {
                state = .loading
            }
            
            let result = await loginUseCase.execute(loginInput: loginCredential.toInput())
            switch result {
            case .success(let userProfile):
                await MainActor.run {
                    state = .success(userProfile)
                }
            case .failure(let error):
                await MainActor.run {
                    state = .failure(error.toLocalizedError())
                }
            }
        }
    }
}

private extension Error {
    
    func toLocalizedError() -> Error {
        if let loginError = self as? LoginError {
            return loginError.toLoginLocalizedError()
        } else if let networkError = self as? NetworkError {
            return networkError.toNetworkLocalizedError()
        }
        
        return UnknownLocalizedError()
    }
}
